Validation/MuonHits/src/GEMSimHitMatcher.cc

```diff
@@ -0,0 +1,360 @@
+#include "Validation/MuonHits/interface/GEMSimHitMatcher.h"
+
+using namespace std;
+
+GEMSimHitMatcher::GEMSimHitMatcher(const edm::ParameterSet& ps,
+                                   edm::ConsumesCollector&& iC)
+    : MuonSimHitMatcher(ps, std::move(iC)) {
+  simHitPSet_ = ps.getParameterSet("gemSimHit");
+  verbose_ = simHitPSet_.getParameter<int>("verbose");
+  simMuOnly_ = simHitPSet_.getParameter<bool>("simMuOnly");
+  discardEleHits_ = simHitPSet_.getParameter<bool>("discardEleHits");
+
+  simHitInput_ = iC.consumes<edm::PSimHitContainer>(
+      simHitPSet_.getParameter<edm::InputTag>("inputTag"));
+}
+
+/// initialize the event
+void GEMSimHitMatcher::init(const edm::Event& iEvent,
+                            const edm::EventSetup& iSetup) {
+  iSetup.get<MuonGeometryRecord>().get(gem_geom_);
+  if (gem_geom_.isValid()) {
+    geometry_ = dynamic_cast<const GEMGeometry*>(&*gem_geom_);
+  } else {
+    hasGeometry_ = false;
+    edm::LogWarning("GEMSimHitMatcher")
+        << "+++ Info: GEM geometry is unavailable. +++\n";
+  }
+  MuonSimHitMatcher::init(iEvent, iSetup);
+}
+
+/// do the matching
+void GEMSimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
+  MuonSimHitMatcher::match(track, vertex);
+
+  if (hasGeometry_) {
+    matchSimHitsToSimTrack(track_ids_, simHits_);
+
+    if (verbose_) {
+      edm::LogInfo("GEMSimHitMatcher")
+          << "nTrackIds " << track_ids_.size() << " nSelectedGEMSimHits "
+          << hits_.size() << endl;
+      edm::LogInfo("GEMSimHitMatcher")
+          << "detids GEM " << detIds(0).size() << endl;
+
+      const auto& gem_ch_ids = detIds();
+      for (const auto& id : gem_ch_ids) {
+        const auto& gem_simhits = MuonSimHitMatcher::hitsInDetId(id);
+        const auto& gem_simhits_gp = simHitsMeanPosition(gem_simhits);
+        edm::LogInfo("GEMSimHitMatcher")
+            << "gemchid " << GEMDetId(id) << ": nHits " << gem_simhits.size()
+            << " phi " << gem_simhits_gp.phi() << " nCh "
+            << chamber_to_hits_[id].size() << endl;
+        const auto& strips = hitStripsInDetId(id);
+        edm::LogInfo("GEMSimHitMatcher") << "nStrip " << strips.size() << endl;
+        edm::LogInfo("GEMSimHitMatcher") << "strips : ";
+        std::copy(strips.begin(), strips.end(),
+                  ostream_iterator<int>(cout, " "));
```

Comment:
This will also write to `cout` which is forbidden. Could you please substitute this call with `MessageLogger` as well?